modules/querier/querier.go

```diff
@@ -184,36 +184,44 @@ func (q *Querier) FindTraceByID(ctx context.Context, req *tempopb.TraceByIDReque
-				span.LogFields(ot_log.String("msg", "combined trace protos"),
+				span.LogFields(ot_log.String("msg", "combined trace protos from ingesters"),
 					ot_log.Int("spansCountA", spanCountA),
 					ot_log.Int("spansCountB", spanCountB),
 					ot_log.Int("spansCountTotal", spanCountTotal))
 			}
 		}
+
+		span.LogFields(ot_log.String("msg", "done searching ingesters"), ot_log.Bool("found", completeTrace != nil))
 	}
 
-	// if the ingester didn't have it check the store.
-	if completeTrace == nil {
-		foundBytes, metrics, err := q.store.Find(opentracing.ContextWithSpan(ctx, span), userID, req.TraceID, req.BlockStart, req.BlockEnd)
-		if err != nil {
-			return nil, errors.Wrap(err, "error querying store in Querier.FindTraceByID")
-		}
+	span.LogFields(ot_log.String("msg", "searching store"))
+	partialTraces, metrics, err := q.store.Find(opentracing.ContextWithSpan(ctx, span), userID, req.TraceID, req.BlockStart, req.BlockEnd)
+	if err != nil {
+		return nil, errors.Wrap(err, "error querying store in Querier.FindTraceByID")
+	}
 
-		out := &tempopb.Trace{}
-		err = proto.Unmarshal(foundBytes, out)
+	span.LogFields(ot_log.String("msg", "done searching store"))
+	metricQueryReads.WithLabelValues("bloom").Observe(float64(metrics.BloomFilterReads.Load()))
+	metricQueryBytesRead.WithLabelValues("bloom").Observe(float64(metrics.BloomFilterBytesRead.Load()))
+	metricQueryReads.WithLabelValues("index").Observe(float64(metrics.IndexReads.Load()))
+	metricQueryBytesRead.WithLabelValues("index").Observe(float64(metrics.IndexBytesRead.Load()))
+	metricQueryReads.WithLabelValues("block").Observe(float64(metrics.BlockReads.Load()))
+	metricQueryBytesRead.WithLabelValues("block").Observe(float64(metrics.BlockBytesRead.Load()))
+
+	// combine partialTraces with completeTrace
+	var spanCountA, spanCountB, spanCountTotal int
+	for _, partialTrace := range partialTraces {
+		storeTrace := &tempopb.Trace{}
+		err = proto.Unmarshal(partialTrace, storeTrace)
 		if err != nil {
 			return nil, err
 		}
-
-		span.LogFields(ot_log.String("msg", "found backend trace"), ot_log.Int("len", len(foundBytes)))
-		completeTrace = out
-		metricQueryReads.WithLabelValues("bloom").Observe(float64(metrics.BloomFilterReads.Load()))
-		metricQueryBytesRead.WithLabelValues("bloom").Observe(float64(metrics.BloomFilterBytesRead.Load()))
-		metricQueryReads.WithLabelValues("index").Observe(float64(metrics.IndexReads.Load()))
-		metricQueryBytesRead.WithLabelValues("index").Observe(float64(metrics.IndexBytesRead.Load()))
-		metricQueryReads.WithLabelValues("block").Observe(float64(metrics.BlockReads.Load()))
-		metricQueryBytesRead.WithLabelValues("block").Observe(float64(metrics.BlockBytesRead.Load()))
+		completeTrace, spanCountA, spanCountB, spanCountTotal = tempo_util.CombineTraceProtos(completeTrace, storeTrace)
 	}
+	span.LogFields(ot_log.String("msg", "combined trace protos from ingesters and store"),
```

Comment:
since we're iterating through a bunch of partial traces and combining them, this spansCountA/B/Total will only be the final iteration which isn't very useful.  I do like a span tag here to give information about the "size" of the combined trace.

maybe just do span total?  or bytes total by adding up the buffer lengths?

Reply:
The store returns marshalled traces but ingester client marshals it into trace object. I wouldn't want to add a marshalling step just to get this span tag, so going with total number of spans.